### PR TITLE
 JSR223: Extended for working with javascript. #2801

### PR DESCRIPTION
**Pull request by lewie**

### CODE REVIEW

**Comment by smerschjohann** [comment]
Looks good to me except the Mysql stuff. I would also like to see "pe" as a shorthand for PersistenceExtensions. Maybe add both?

And please move your author line below mine.

Last but not least: Please squash your commits into a single one, that it can be merged into the master branch

**Comment by lewie** [comment]
After a few confused Git operations.
Et voilà

**Comment by teichsta** [comment]
with your new knowledge about git, could you please squash these commits as well? Thanks, Thomas E.-E.

**Comment by lewie** [comment]
@teichsta: Updated, 1 Commit

@smerschjohann: 
```
Shorthand:
ir	Itemregistry
be	Busevent
pe	PersistenceExtensions
oh	Openhab

Imports especially for JavaScript:
transform			Transformation Action
getItem 			ItemRegistry.getItem
postUpdate 			BusEvent.postUpdate
sendCommand 		BusEvent.sendCommand
FileUtils 			org.apache.commons.io.FileUtils
File 				java.io.File
```

**Comment by smerschjohann** [comment]
looks good to me, except the little thing I mentioned as line comment.

I prefer this PR, as it supports one additional script engine.

**Comment by lewie** [comment]
@smerschjohann: Ys, siorry, better now?

**Comment by lewie** [comment]
@watou: Done curly braces. Sorry my git push deletd your comment.

**Comment by watou** [comment]
Thanks, and thanks for adding JavaScript!

**Comment by lewie** [comment]
@watou: Netxt step, I can work in a Browser (ACE-Editor). For this, doing Tests at the moment, JS seams to be minimum as fast as xText. :-)

**Comment by lewie** [comment]
Updated because found a Bug: Have to apply all Classes in Script.java before it can be imported by nashorn.

File: bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java [hunk 1]
```diff
@@ -52,13 +60,15 @@
  * Script-Engine to allow interoperability with openHAB.
  * 
  * @author Simon Merschjohann
+ * @author Helmut Lehmeyer
  * @since 1.7.0
  */
+//@SuppressWarnings("unused")
```
Comment (teichsta) [inline bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java]: please remove commented code

File: bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java [hunk 2]
```diff
@@ -68,52 +78,133 @@ public Script(ScriptManager scriptManager, File file) throws FileNotFoundExcepti
 	}
 
 	public void loadScript(File file) throws FileNotFoundException, ScriptException, NoSuchMethodException {
-		logger.info("Loading Script " + file.getName());
+		logger.info("loadScript Loading Script " + file.getName());
```
Comment (teichsta) [inline bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java]: this addition doesn't seem to make sense

File: bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java [hunk 3]
```diff
@@ -68,52 +78,133 @@ public Script(ScriptManager scriptManager, File file) throws FileNotFoundExcepti
...
+						
+						+"ohEngine				= 'javascript';\n"
+						
+						//Helper Functions and Libs eventually later a lib Folder for default (Auto) loaded Libraries
+						//Bas64: https://gist.github.com/ncerminara/11257943
+						//+"load('configurations/scripts/jslib/b64.js');\n"
+						);
+			} catch (ScriptException e) {
+				logger.error("ScriptException in initializeSciptGlobals while importing default-classes: ", e);
+			}
```
Comment (teichsta) [inline bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java]: please extract into it's own initMethod for the sake of readability of this general init method

File: bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java [hunk 4]
```diff
@@ -68,52 +78,133 @@ public Script(ScriptManager scriptManager, File file) throws FileNotFoundExcepti
...
+			engine.put("HSBType", 				HSBType.class);
+			engine.put("IncreaseDecreaseType", 	IncreaseDecreaseType.class);
+			engine.put("OnOffType", 			OnOffType.class);
+			engine.put("OpenClosedType", 		OpenClosedType.class);
+			engine.put("PercentType", 			PercentType.class);
+			engine.put("PointType", 			PointType.class);
+			engine.put("StopMoveType", 			StopMoveType.class);
+			engine.put("UpDownType", 			UpDownType.class);
+			engine.put("StringType", 			StringType.class);
+		}
```
Comment (teichsta) [inline bundles/core/org.openhab.core.jsr223/src/main/java/org/openhab/core/jsr223/internal/engine/scriptmanager/Script.java]: please extract into it's own initMethod for the sake of readability of this general init method

File: distribution/openhabhome/configurations/scripts/jsr223_demo.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    }
+};
+
+
+
+function getRules(){
+	return new RuleSet([TestRule, ActionTest, FileOperations]);
+}
+
```
Comment (teichsta) [inline distribution/openhabhome/configurations/scripts/jsr223_demo.js]: please remove this sample file and rather add it to the AddOn itself or (even better) enhance the Wiki page https://github.com/openhab/openhab/wiki/Jsr223-Script-Engine

File: bundles/core/org.openhab.core.jsr223/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -6,8 +6,12 @@ Bundle-Vendor: openHAB.org
 Bundle-Version: 1.8.0.qualifier
 Bundle-SymbolicName: org.openhab.core.jsr223;singleton:=true
 Bundle-ActivationPolicy: lazy
-Require-Bundle: org.openhab.model.item
+Require-Bundle: org.openhab.model.item,
+ org.openhab.core.persistence,
+ org.openhab.model.script
```
Comment (teichsta) [inline bundles/core/org.openhab.core.jsr223/META-INF/MANIFEST.MF]: please avoid using require-bundle deps in favour of import-package statements (see also https://github.com/openhab/openhab/wiki/How-To-Contribute#OSGi Bundles)

File: distribution/openhabhome/start.sh [hunk 7]
```diff
@@ -14,6 +14,7 @@ cp=$(find $eclipsehome -name "org.eclipse.equinox.launcher_*.jar" | sort | tail
 
 echo Launching the openHAB runtime...
 java \
+	-Dorg.osgi.framework.bundle.parent=ext \
```
Comment (teichsta) [inline distribution/openhabhome/start.sh]: hmm … changing the ClassLoader behaviour of the whole openHAB instance feels like a fundamental change. Any chance to make this change more focussed (see https://www.eclipse.org/forums/index.php/t/217277/)?

**Comment by teichsta** [comment]
Hi @lewie thanks for this contribution and for squashing the commits. Nevertheless i've added some review comments inline. Best, Thomas E.-E.

**Comment by lewie** [comment]
Hi @teichsta, 
squashed the commits. ;-)

ClassLoader behaviour:
After some hours learning much about class loaders in Java. It seems, Java 9 is already casting shadows, there ClassLoader behavior "extension" will be standard. 
Why "-Dorg.osgi.framework.bundle.parent=ext" is needed: Nashorn does not join into the OSGI classloader hierarchy, it is at the level of a bundle.  In addition, it has a little different security model for loading Java classes in Javascript. In short, I have not found an adequate solution through ClassLoader manipulations. Now, I check the used ClassLoader, if false, gets a Warning. The explanation I write to the wiki.

Best, Helmut

**Comment by teichsta** [comment]
great, thanks a lot @lewie!

**Pull request merged**